File: Gordon_Grant_FSharp/Checkout.fs
module Checkout

type Product = Apple | Beans | Coke | Deodorant

let checkout (items:Product list) (prices:Map<Product,float>) (discounts:Map<Product list,float>) = 
    
    let (undiscountedItems, discountsTotal) =
        discounts 
        |> Map.fold (fun agg key value -> 
                let remainingItems = (fst agg)
                let discountProducts = key |> List.toArray
                let discountedProductCount = key |> List.length

                let discountFound = 
                    (remainingItems |> Array.length >= discountedProductCount)
                    && (remainingItems.[0..discountedProductCount-1] = discountProducts)

                match discountFound with
                | true -> (remainingItems.[discountedProductCount..]),(snd agg + value)
                | _ -> agg)
            (items |> List.toArray,0.0)

    discountsTotal + (undiscountedItems |> Array.sumBy (fun item -> prices.[item]))

File: Phil_Chris_Elixir/lib/checkout.ex
defmodule Checkout do
	def price(items) do
		Enum.sort(items) |> total(0)
	end

	def item([:apple,:apple,:apple,:apple|shopping]) do
		{1.00, shopping}
	end
	def item([:deodorant, :deodorant|shopping]) do
		{4.50, shopping}
	end
	def item([:apple|shopping]) do
		{0.30, shopping}
	end
	def item([:beans|shopping]) do
		{0.50, shopping}
	end
	def item([:coke|shopping]) do
		{1.80, shopping}
	end
	def item([:deodorant|shopping]) do
		{2.50, shopping}
	end


	def total([], acc) do
		acc
	end
	def total(shopping_list, acc) do
		{price, shopping} = item(shopping_list)
		total(shopping, acc + price)
	end
end
